Mitigate support for OpenAPI 3.1

SergeyMenshykh opened the pull request:
### Motivation and Context

This PR adds support for OpenAPI 3.1 spec to OpenAPI skills functionality that may be useful in scenarios when OpenAPI 3.1 skills are imported directly from OpenAPI documents or indirectly through ChatGPT plugins.

### Description

To support the OpenAPI 3.1 specification, the OpenApiDocumentParser class performs the following actions:
- Adapts/converts YAML content to JSON for further processing using System.Text.Json API.
- Tries to find the 'openapi' property.
- Downgrade the property value to latest supported version - 3.0.1 if the version in the document is higher.
- Passes the updated JSON OpenAPI document to Microsoft.OpenAPI.NET library for parsing.

Other small changes introduced in this PR:
- Cancellation toke is added to all ImportChatGptPluginSkill* and ImportOpenApiSkill* methods.
- All the ImportChatGptPluginSkill* and ImportOpenApiSkill* methods are made asynchronous.
- ReadAsync method of OpenApiReader class is used Read one.


### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows SK Contribution Guidelines (https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md)
- [x] The code follows the .NET coding conventions (https://learn.microsoft.com/dotnet/csharp/fundamentals/coding-style/coding-conventions) verified with `dotnet format`
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

<!-- Thank you for your contribution to the semantic-kernel repo! -->

lemillermicrosoft commented on dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/Extensions/KernelOpenApiExtensions.cs:
```diff
@@ -65,7 +72,7 @@ public static class KernelOpenApiExtensions
             }
 
-            return kernel.RegisterOpenApiSkill(stream, skillName, authCallback);
+            return await kernel.RegisterOpenApiSkillAsync(stream, skillName, authCallback, cancellationToken);
```
Curious if the await here has an impact, is it not better to just return the task when possible?

SergeyMenshykh replied:
On the one hand, there are inefficiencies introduced by compiler that generates a state-machine code for async/await methods. On the other hand, there are a few pitfalls you need to make sure not fall into when skipping async/await keywords. I’ll follow whichever adopted approach, but I would be more inclined to use async/await to save my dev time by not thinking/analyzing consequences of omitting them, up to the point when it becomes an issue, if at all, and only in this case will think about optimization.

lemillermicrosoft replied:
@shawncal in case we have a team preference.

shawncal replied:
Where there's only one await and it's on the final statement, I prefer to remove async and the await, just return the Task directly.

lemillermicrosoft replied:
Sounds good, so if there is other await in the method already and we can't remove the `async` then it's fine to `return await` . Thanks!

lemillermicrosoft reviewed (dismissed):
Overall LGTM.

shawncal reviewed (commented):
Just 2 comments:
 - async/await thing mentioned below
 - where we are adding awaits, please add start adding .ConfigureAwait(false) as a general good practice (we'll need to make this change in the whole kernel project at some point)

SergeyMenshykh commented:
> Just 2 comments:
> 
> * async/await thing mentioned below
> * where we are adding awaits, please add start adding .ConfigureAwait(false) as a general good practice (we'll need to make this change in the whole kernel project at some point)

Removed the awaits where it was possible and added ConfigureAwait(false) to all awaits in OpenAPI sills functionality.

adrianwyatt commented on dotnet/src/SemanticKernel.Skills/Skills.OpenAPI/OpenApi/OpenApiDocumentParser.cs:
```diff
@@ -21,21 +26,88 @@ namespace Microsoft.SemanticKernel.Skills.OpenAPI.OpenApi;
+    public async Task<IList<RestApiOperation>> ParseAsync(Stream stream, CancellationToken cancellationToken = default)
     {
-        var openApiDocument = new OpenApiStreamReader().Read(stream, out var diagnostic);
+        var jsonObject = await this.DowngradeDocumentVersionToSuportedOneAsync(stream, cancellationToken).ConfigureAwait(false);
```
Could you also update the title of the PR to replace "Support" with "Mitigate" since we don't *technically* support true OpenAPI 3.1. Of course, let me know if I'm misunderstanding.
